strawberry/schema/schema.py

```diff
@@ -214,15 +214,57 @@ class Query:
             formatted_errors = "\n\n".join(f"❌ {error.message}" for error in errors)
             raise ValueError(f"Invalid Schema. Errors:\n\n{formatted_errors}")
 
-    def get_extensions(
-        self, sync: bool = False
-    ) -> List[Union[Type[SchemaExtension], SchemaExtension]]:
-        extensions = list(self.extensions)
+    def get_extensions(self, sync: bool = False) -> List[SchemaExtension]:
+        if not self._cached_extensions:
+            extensions = []
+            if self.directives:
+                extensions = [
+                    *self.extensions,
+                    DirectivesExtensionSync if sync else DirectivesExtension,
+                ]
+            extensions.extend(self.extensions)
+            self._cached_extensions = [
+                ext if isinstance(ext, SchemaExtension) else ext(execution_context=None)
+                for ext in extensions
+            ]
+        return self._cached_extensions
+
+    def create_extensions_runner(
+        self, execution_context: ExecutionContext, extensions: list[SchemaExtension]
+    ) -> SchemaExtensionsRunner:
+        return SchemaExtensionsRunner(
+            execution_context=execution_context,
+            extensions=extensions,
+        )
 
-        if self.directives:
-            extensions.append(DirectivesExtensionSync if sync else DirectivesExtension)
+    def _get_middleware_manager(
```

Comment:
This one is kinda trickier because `extensions` are not just a boolean argument. I think the current approuch suffice.
LMK if thats a blocker, I'll think of something better.